data/shared/src/main/scala/org/ergoplatform/HeaderWithoutPow.scala

```diff
@@ -0,0 +1,139 @@
+                       val stateRoot: ADDigest, //33 bytes! extra byte with tree height here!
+                       val transactionsRoot: Digest32, // 32 bytes
+                       val timestamp: Long,
+                       val nBits: Long, //actually it is unsigned int
+                       val height: Int,
+                       val extensionRoot: Digest32,
+                       val votes: Array[Byte], //3 bytes
+                       val unparsedBytes: Array[Byte]) {
+  def toHeader(powSolution: AutolykosSolution, bytes: Array[Byte]): ErgoHeader =
+    ErgoHeader(version, parentId, ADProofsRoot, stateRoot, transactionsRoot, timestamp,
+      nBits, height, extensionRoot, powSolution, votes, unparsedBytes, bytes)
+}
+
+object HeaderWithoutPow {
+
+  def apply(version: Byte, parentId: ModifierId, ADProofsRoot: Digest32, stateRoot: ADDigest,
+            transactionsRoot: Digest32, timestamp: Long, nBits: Long, height: Int,
+            extensionRoot: Digest32, votes: Array[Byte], unparsedBytes: Array[Byte]): HeaderWithoutPow = {
+    new HeaderWithoutPow(version, parentId, ADProofsRoot, stateRoot, transactionsRoot, timestamp,
+      nBits, height, extensionRoot, votes, unparsedBytes)
+  }
+
+}
+
+object HeaderWithoutPowSerializer extends SigmaSerializer[HeaderWithoutPow, HeaderWithoutPow] {
+
+  override def serialize(h: HeaderWithoutPow, w: SigmaByteWriter): Unit = {
+    w.put(h.version)
+    w.putBytes(idToBytes(h.parentId))
+    w.putBytes(h.ADProofsRoot)
+    w.putBytes(h.transactionsRoot)
+    w.putBytes(h.stateRoot)
+    w.putULong(h.timestamp)
+    w.putBytes(h.extensionRoot)
+    DifficultySerializer.serialize(h.nBits, w)
+    w.putUInt(h.height.toLong)
+    w.putBytes(h.votes)
+
+    // For block version >= 2, this new byte encodes length of possible new fields.
+    // Set to 0 for now, so no new fields.
+    if (h.version > HeaderVersion.InitialVersion) {
+      w.putUByte(h.unparsedBytes.length)
+      w.putBytes(h.unparsedBytes)
+    }
+  }
+
+  override def parse(r: SigmaByteReader): HeaderWithoutPow = {
+    val version = r.getByte()
+    val parentId = bytesToId(r.getBytes(32))
+    val ADProofsRoot = Digest32 @@ r.getBytes(32)
+    val transactionsRoot = Digest32 @@ r.getBytes(32)
+    val stateRoot = ADDigest @@ r.getBytes(33)
+    val timestamp = r.getULong()
+    val extensionHash = Digest32 @@ r.getBytes(32)
+    val nBits = DifficultySerializer.parse(r)
+    val height = r.getUInt().toIntExact
+    val votes = r.getBytes(3)
+
+    // For block version >= 2, a new byte encodes length of possible new fields.
+    // If this byte > 0, we read new fields but do nothing, as semantics of the fields is not known.
+    val unparsedBytes = if (version > HeaderVersion.InitialVersion) {
+      val newFieldsSize = r.getUByte()
+      if (newFieldsSize > 0 && version > HeaderVersion.Interpreter60Version) {
+        // new bytes could be added only for block version >= 5
+        r.getBytes(newFieldsSize)
```

Comment:
??? 
When nodes with V7 activated will start to produce V7 headers, v6 (actually >= 5.0.22) nodes will read and store unparsed bytes. 

see https://github.com/ergoplatform/ergo/pull/2151

Reply:
Still don't understand why the parsing was changed. 
Currenty the byte is reserved and is expected to be zero, but I don't see where it is enforced.
I can put 2 and then put 2 extra bytes and it will be valid v5.0 header and will be parsed.
So, the only missing part is storing what was parsed in the HeaderWithoutPow instance and update serialization to save extra bytes.
The name `unparsedBytes` is somewhat misleading, because they are literally parsed, but obscure/opaque.
So, `extraBytes` would be better name, and then in future versions they can receive some meaning without stopping being "extra".

Reply:
How POW is related to serialization? Header can be serializable, but not valid for protocol, so validation is a layer on top of serialization.

Reply:
PoW is generated over HeaderWithoutPoW bytes.

Reply:
If version == 6, and newFieldsSize > 0, then 
- v5.0 node will parse the header, read all the bytes, but will not save them in the header instance.
- v6.0 node will not read the bytes leaving reader in an invalid state which will lead to failure later
Thus, we allow newFieldsSize > 0 in v5.0 and forbid it in v6.0, right?

Reply:
V 5.0 will parse the header, but PoW check will be failed after. So consensus-wise both 5.0 and 6.0 will reject the header, only with different errors.